Mathlib/Probability/Bernoulli.lean

```diff
@@ -0,0 +1,208 @@
+  obtain rfl | hμ := eq_zero_or_isProbabilityMeasure μ
+  · simp
+  calc
+    _ = μ[X ^ 2] - μ[X] ^ 2 := variance_eq_sub <| .of_bound hXmeas.aestronglyMeasurable 1 <| by
+        filter_upwards [hX]; rintro ω (hω | hω) <;> simp [hω]
+    _ = μ[X] - μ[X] ^ 2 := by
+      congr! 1
+      exact integral_congr_ae <| by filter_upwards [hX]; rintro ω (hω | hω) <;> simp [hω]
+    _ = μ.real {ω | X ω = 0} * μ.real {ω | X ω = 1} := by
+      rw [sq, ← one_sub_mul, integral_of_ae_eq_zero_or_one hXmeas.aemeasurable hX]
+      congr
+      rw [← ENNReal.toReal_one, ← ENNReal.toReal_sub_of_le, ← prob_compl_eq_one_sub]
+      · congr 1
+        refine measure_congr ?_
+        filter_upwards [hX]
+        -- FIXME: The following change is due to the measure theory library abusing the defeq
+        -- `Set Ω = (Ω → Prop)`
+        change ∀ ω _, (_ ≠ _) = (_ = _)
+        rintro ω (hω | hω) <;> simp [hω]
+      · exact hXmeas <| .singleton _
+      · exact prob_le_one
+      · simp
+
+/-- If a random variable is ae equal to `0` or `1`, then its conditional variance is the product of
+the conditional probabilities that it's equal to `0` and that it's equal to `1`. -/
+lemma condVar_of_ae_eq_zero_or_one {m₀ : MeasurableSpace Ω} (hm : m ≤ m₀) {μ : Measure[m₀] Ω}
+    [IsFiniteMeasure μ] (hXmeas : AEMeasurable[m₀] X μ) (hX : ∀ᵐ ω ∂μ, X ω = 0 ∨ X ω = 1) :
+    Var[X; μ | m] =ᵐ[μ] μ[X | m] * μ[1 - X | m] := by
+  wlog hXmeas : Measurable[m₀] X
+  · obtain ⟨Y, hYmeas, hXY⟩ := ‹AEMeasurable[m₀] X μ›
+    calc
+      Var[X; μ | m]
+      _ =ᵐ[μ] Var[Y; μ | m] := condVar_congr_ae hXY
+      _ =ᵐ[μ] μ[Y | m] * μ[1 - Y | m] := by
+        refine this hm hYmeas.aemeasurable ?_ hYmeas
+        filter_upwards [hX, hXY] with ω hXω hXYω
+        simp [hXω, ← hXYω]
+      _ =ᵐ[μ] μ[X | m] * μ[1 - X | m] := by
+        refine .mul ?_ ?_ <;>
+          exact condExp_congr_ae <| by filter_upwards [hXY] with ω hω; simp [hω]
+  calc
+    _ =ᵐ[μ] μ[X ^ 2 | m] - μ[X | m] ^ 2 :=
+      condVar_ae_eq_condExp_sq_sub_sq_condExp hm <| .of_bound hXmeas.aestronglyMeasurable 1 <| by
+        filter_upwards [hX]; rintro ω (hω | hω) <;> simp [hω]
+    _ =ᵐ[μ] μ[X | m] - μ[X | m] ^ 2 := by
+      refine .sub ?_ ae_eq_rfl
+      exact condExp_congr_ae <| by filter_upwards [hX]; rintro ω (hω | hω) <;> simp [hω]
+    _ =ᵐ[μ] μ[X | m] * μ[1 - X | m] := by
+      rw [sq, ← one_sub_mul, mul_comm]
+      refine .mul ae_eq_rfl ?_
+      calc
+        1 - μ[X | m]
+        _ = μ[1 | m] - μ[X | m] := by simp [Pi.one_def, hm]
+        _ =ᵐ[μ] μ[1 - X | m] := by
+          refine (condExp_sub (integrable_const _)
+            (.of_bound (C := 1) hXmeas.aestronglyMeasurable ?_) _).symm
+          filter_upwards [hX]
+          rintro ω (hω | hω) <;> simp [hω]
+
+/-! ### Predicate for a measure to be Bernoulli -/
+
+/-- The Bernoulli probability distribution with parameter `p`. -/
+noncomputable def bernoulli (p : ℝ≥0) : Measure ℝ := (1 - p) • .dirac 0 + (1 - (1 - p)) • .dirac 1
```

Comment:
Please add a comment in the docstring to explain the use of `1 - (1 - p)` in the def.